repo/migrations/Migration018.go

```diff
@@ -0,0 +1,233 @@
+// We've also removed the resolvers config as we aren't using that any more and added
+// an IpnsExtra config which contains fields we previously patched into the IPNS config.
+type Migration018 struct{}
+
+func (Migration018) Up(repoPath string, dbPassword string, testnet bool) error {
+	configFile, err := ioutil.ReadFile(path.Join(repoPath, "config"))
+	if err != nil {
+		return err
+	}
+	var cfgIface interface{}
+	json.Unmarshal(configFile, &cfgIface)
+	cfg, ok := cfgIface.(map[string]interface{})
+	if !ok {
+		return errors.New("Invalid config file")
+	}
+
+	// Update API address to use nil instead of ""
+	addrsIface, ok := cfg["Addresses"]
+	if !ok {
+		return errors.New("missing addresses config")
+	}
+	addrs, ok := addrsIface.(map[string]interface{})
+	if !ok {
+		return errors.New("error parsing addresses config")
+	}
+	apiAddr, ok := addrs["API"]
+	if !ok {
+		return errors.New("missing API address")
+	}
+	apiAddrStr, ok := apiAddr.(string)
+	if !ok {
+		return errors.New("error parsing API addr")
+	}
+	if apiAddrStr == "" {
+		addrs["API"] = nil
+	}
+
+	// Delete resolvers
+	delete(cfg, "Resolvers")
+
+	// Add APICommands to gateway
+	gatewayIface, ok := cfg["Gateway"]
+	if !ok {
+		return errors.New("missing gateway config")
+	}
+	gatewayCfg, ok := gatewayIface.(map[string]interface{})
+	if !ok {
+		return errors.New("error parsing gateway config")
+	}
+	gatewayCfg["APICommands"] = nil
+
+	// Modify IPNS config
+	ipnsIface, ok := cfg["Ipns"]
+	if !ok {
+		return errors.New("missing ipns config")
+	}
+	ipnsCfg, ok := ipnsIface.(map[string]interface{})
+	if !ok {
+		return errors.New("error parsing ipns config")
+	}
+	ipnsCfg["RecordLifetime"] = "168h"
+	delete(ipnsCfg, "BackUpAPI")
+	delete(ipnsCfg, "UsePersistentCache")
+	delete(ipnsCfg, "QuerySize")
+
+	cfg["Ipns"] = ipnsCfg
+
+	// Add Pubsub config
+	cfg["Pubsub"] = map[string]interface{}{
+		"DisableSigning":              false,
+		"Router":                      "",
+		"StrictSignatureVerification": false,
+	}
+
+	// Modify Experimental config
+	expIface, ok := cfg["Experimental"]
+	if !ok {
+		return errors.New("missing experimental config")
+	}
+	expCfg, ok := expIface.(map[string]interface{})
+	if !ok {
+		return errors.New("error parsing experimental config")
+	}
+	expCfg["P2pHttpProxy"] = false
+	expCfg["QUIC"] = false
+	expCfg["UrlstoreEnabled"] = false
+
+	cfg["Experimental"] = expCfg
+
+	// Add IPNSExtra config
+	cfg["IpnsExtra"] = map[string]interface{}{
+		"DHTQuorumSize": 1,
+		"FallbackAPI":   "https://gateway.ob1.io",
+	}
+
+	out, err := json.MarshalIndent(cfg, "", "   ")
+	if err != nil {
+		return err
+	}
+	f, err := os.Create(path.Join(repoPath, "config"))
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(out)
+	if err != nil {
+		return err
+	}
+	f.Close()
+	if err := writeRepoVer(repoPath, 19); err != nil {
+		return fmt.Errorf("bumping repover to 19: %s", err.Error())
+	}
+	if err := writeIPFSVer(repoPath, 7); err != nil {
```

Comment:
I expected this version number to be managed internally by the IPFS repo migration handling. Any reason to manage both version numbers separately?

Reply:
IPFS has a separate migration binary ifaik. It doesnt migrate automatically. It's up to the user to run the migration manually. So for us we have to manually migrate the repo.